### PR TITLE
Update index mappings when ccr restore complete

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This is related to #35975. When the shard restore process is complete,
the index mappings need to be updated to ensure that the data in the
files restores is compatible with the follower mappings. This commit
implements a mapping update as the final step in a shard restore.

### CODE REVIEW

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java [hunk 1]
```diff
@@ -261,13 +267,33 @@ public void restoreShard(IndexShard indexShard, SnapshotId snapshotId, Version v
...
         closeSession(remoteClient, nodeId, sessionUUID);
+        maybeUpdateMappings(client, remoteClient, leaderIndex, indexShard.indexSettings());
     }
 
     @Override
     public IndexShardSnapshotStatus getShardSnapshotStatus(SnapshotId snapshotId, Version version, IndexId indexId, ShardId leaderShardId) {
         throw new UnsupportedOperationException("Unsupported for repository of type: " + TYPE);
     }
 
+    private void maybeUpdateMappings(Client localClient, Client remoteClient, Index leaderIndex, IndexSettings followerIndexSettings) {
```
Comment (ywelsch) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java]: @martijnvg would it make sense to share this code with the one in ShardFollowTasksExecutor?

**Review by ywelsch — APPROVED** [review]
@martijnvg can you have a look as well please?

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java (hunk 1 above)
Comment (martijnvg) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java]: Conceptually this code is doing the same as the update mapping code in shard follow task, but the update mapping code is tightly coupled with `ShardFollowTasksExecutor` and `ShardFollowNodeTask`. Also the mapping update code is asynchronous over there and here it is synchronous.

What I think is possible, is that factory methods for both `ClusterStateRequest` and `PutMappingRequest` are introduced that is then used here and in `ShardFollowTasksExecutor`.

**Comment by tbrooks8** [comment]
@martijnvg @ywelsch I made some changes to share the request creation.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrRequests.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import org.elasticsearch.action.admin.cluster.state.ClusterStateRequest;
+import org.elasticsearch.action.admin.indices.mapping.put.PutMappingRequest;
+import org.elasticsearch.cluster.metadata.MappingMetaData;
+import org.elasticsearch.common.xcontent.XContentType;
+
+public final class CcrRequests {
+
+    private CcrRequests() {}
+
+    public static ClusterStateRequest clusterStateRequest(String leaderIndex) {
```
Comment (ywelsch) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrRequests.java]: perhaps `metaDataRequest`?

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java [hunk 3]
```diff
@@ -116,15 +116,9 @@ public SnapshotInfo getSnapshotInfo(SnapshotId snapshotId) {
...
-        ClusterStateResponse response = remoteClient
-            .admin()
-            .cluster()
-            .prepareState()
-            .clear()
-            .setMetaData(true)
-            .setIndices("dummy_index_name") // We set a single dummy index name to avoid fetching all the index data
-            .get();
-        return response.getState().metaData();
+        ClusterStateRequest clusterStateRequest = CcrRequests.clusterStateRequest("dummy_index_name");
```
Comment (ywelsch) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java]: you lost get comment about "dummy_index_name"

**Review by ywelsch — CHANGES_REQUESTED** [review]

**Review by ywelsch — APPROVED** [review]
LGTM

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrRequests.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ccr.action;
+
+import org.elasticsearch.action.admin.cluster.state.ClusterStateRequest;
+import org.elasticsearch.action.admin.indices.mapping.put.PutMappingRequest;
+import org.elasticsearch.cluster.metadata.MappingMetaData;
+import org.elasticsearch.common.xcontent.XContentType;
+
+public final class CcrRequests {
```
Comment (martijnvg) [inline x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/CcrRequests.java]: 👍

**Pull request merged**